components/supervisor/pkg/supervisor/notification.go

```diff
@@ -0,0 +1,161 @@
+}
+
+// Sends a notification to the user
+func (srv *NotificationService) Notify(ctx context.Context, req *api.NotifyRequest) (*api.NotifyResponse, error) {
+	if srv.nextNotificationId >= maxPendingNotifications {
+		return nil, status.Error(codes.Internal, "Max number of pending notifications exceeded")
+	}
+	var channel = srv.notifySubscribers(req)
+	if len(req.Actions) == 0 {
+		return &api.NotifyResponse{}, nil
+	}
+	select {
+	case resp := <-channel:
+		return resp, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (srv *NotificationService) notifySubscribers(req *api.NotifyRequest) chan *api.NotifyResponse {
+	srv.mutex.Lock()
+	defer srv.mutex.Unlock()
+	var (
+		requestId = srv.nextNotificationId
+		message   = &api.SubscribeResult{
+			RequestId: requestId,
+			Request:   req,
+		}
+		staleSubscribers = []int{}
+	)
+	srv.nextNotificationId++
+	for i, subscriber := range srv.subscribers {
+		// TODO: why do I have to dereference here?
+		var err = subscriber.Send(message)
+		if err != nil {
+			staleSubscribers = append(staleSubscribers, i)
+		}
+	}
+	srv.removeSubscribers(staleSubscribers)
+	var channel = make(chan *api.NotifyResponse)
+	srv.pendingNotifications[requestId] = &pendingNotification{
+		message:         message,
+		responseChannel: channel,
+	}
+	return channel
+}
+
+// subscribes to notifications that are sent to the supervisor
+func (srv *NotificationService) Subscribe(req *api.SubscribeRequest, resp api.NotificationService_SubscribeServer) error {
+	srv.mutex.Lock()
+	defer srv.mutex.Unlock()
+	for id, pending := range srv.pendingNotifications {
+		var err = resp.Send(pending.message)
+		if err != nil {
+			return status.Errorf(codes.FailedPrecondition, "Cannot subscribe new subscriber as sending pending notification failed. %s", err)
+		}
+		if len(pending.message.Request.Actions) == 0 {
+			delete(srv.pendingNotifications, id)
+		}
+	}
+	srv.subscribers = append(srv.subscribers, resp)
+	return nil
+}
+
+// reports user actions as response to a notification request
+func (srv *NotificationService) Respond(ctx context.Context, req *api.RespondRequest) (*api.RespondResult, error) {
+	srv.mutex.Lock()
+	defer srv.mutex.Unlock()
+	var pending = srv.pendingNotifications[req.RequestId]
+	if pending == nil {
+		log.Log.WithField("requestId", req.RequestId).Info("Invalid or late response to notification")
+		return nil, status.Errorf(codes.InvalidArgument, "Invalid or late response to notification")
+	}
+	if !isActionAllowed(req.Response.Action, pending.message.Request) {
+		log.Log.WithFields(map[string]interface{}{
+			"Notification": pending.message,
+			"Action":       req.Response.Action,
+		}).Error("Invalid user action on notification")
+		return nil, status.Errorf(codes.InvalidArgument, "Invalid user action on notification")
+	}
+	delete(srv.pendingNotifications, req.RequestId)
+	pending.responseChannel <- req.Response
```

Comment:
This write might block if the context of the original "Notify" was canceled. See comment above.

Reply:
OK